site/app/controllers/admin/ReportController.php

```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
             new WebResponse(
                 ['admin', 'Report'],
                 'showFullGradebook',
-                $grade_file
+                $grade_file,
+                $grade_summaries_last_run
             )
         );
     }
+
+    /**
+     * Generate a custom filename for the downloaded CSV file
+     *
+     * @return string
+     */
+    private function generateCustomFilename() {
+        $course = $this->core->getConfig()->getCourse();
+        $timestamp = date("ymdHis");
+        return "{$course}_rainbow_grades_{$timestamp}.csv";
+    }
+
+
+    /**
+     * Download CSV file for Rainbow Grades
+     * @return void
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile() {
+        // Check if the user has the necessary permissions (e.g., admin access)
+        if (!$this->core->getUser()->accessAdmin()) {
+            $this->core->getOutput()->showError("This account cannot access admin pages");
+        }
+
+        // Path to the CSV file for Rainbow Grades
+        $csvFilePath = '/var/local/submitty/courses/' . $this->core->getConfig()->getTerm() . '/' . $this->core->getConfig()->getCourse() . '/rainbow_grades/output.csv';
+
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
+            header('Pragma: public');
+            header('Content-Length: ' . filesize($csvFilePath));
```

Comment:
[this](https://github.com/Submitty/Submitty/pull/10079#discussion_r1438422010) review comment from Chris still needs to be addressed.

Reply:
https://github.com/Submitty/Submitty/pull/10079#discussion_r1530954806

Reply:
Don't we already have some existing code for file downloads?